### PR TITLE
[BB-9214] Implement pathway progress api

### PR DESCRIPTION
**Pull request by pkulkark**
## Description
This PR adds an endpoint that returns the aggregate progress of the requested user in a learning path. It in turn depends on the completion aggregator API for fetching the completion percentages of the individual courses that make up a learning path.

## Supporting Information
OpenCraft Internal Jira ticket: [BB-9214](https://tasks.opencraft.com/browse/BB-9214)

## Testing Instructions
1. Set up devstack.
2. Clone this PR branch repository to `<devstack root>/src`
3. Enter lms shell: `$ make dev.shell.lms`
4. Install this plugin: `# pip install -e /edx/src/learning-paths-plugin/`
5. Run migrations: `# ./manage.py lms migrate learning_paths`
6. Run completion aggregator migrations: `# ./manage.py lms migrate completion_aggregator`
7. Once the plugin is installed, create a learning path http://localhost:18000/admin/learning_paths/learningpath/ with some test courses and enroll a user into it.
8. Open the learning path and note the id from the address bar. It will be in the form of `admin/learning_paths/learningpath/<id>/change/`.
9. Login to lms as the enrolled user.
10. While being logged in, test the learning path progress using http://localhost:18000/api/v1/learning-paths/1/progress/.
11. Go back to lms and progress through the course(s).
12. Verify that the progress of the learning path is also updated.

### CODE REVIEW

File: README.rst [hunk 1]
```diff
@@ -15,3 +15,45 @@ The code in this repository is licensed under the Not open source unless
...
+
+    Ensure that the **completion aggregator** service is also up to date by running its migrations:
+
+    ```bash
+    ./manage.py lms migrate completion_aggregator
+    ```
+
+5. **Restart LMS**
+
+    Restart the LMS to apply the changes:
```
Comment (Agrendalath) [inline README.rst:57]: Tiny nit: this one seems to be redundant.

File: README.rst [hunk 2]
```diff
@@ -15,3 +15,45 @@ The code in this repository is licensed under the Not open source unless
...
+
+Once these steps are complete, the Learning Paths plugin should be successfully installed and ready to use.
```
Comment (Agrendalath) [inline README.rst:59]: Could you please explain what "ready to use" means? It would be nice to have a "Usage" section in this README.

File: learning_paths/api/v1/views.py [hunk 3]
```diff
@@ -23,3 +25,19 @@ class LearningPathAsProgramViewSet(viewsets.ReadOnlyModelViewSet):
     permission_classes = (IsAuthenticated,)
     serializer_class = LearningPathAsProgramSerializer
     pagination_class = PageNumberPagination
+
+
+def learning_path_progress_view(request, learning_path_id):
```
Comment (Agrendalath) [inline learning_paths/api/v1/views.py:30]: Nit: shouldn't we use the DRF `APIView` and a serializer to keep it consistent with the other view? Also, if we plan to add the view to list multiple users' progress in a learning path, it could make sense to call this something like `LearningPathUserProgressView`.

We should also remember to use `permission_classes = (IsAuthenticated,)`. Currently, this endpoint raises the `AttributeError` when accessed by an anonymous user.

File: learning_paths/api/v1/urls.py [hunk 4]
```diff
@@ -1,12 +1,22 @@
...
 
 router = routers.SimpleRouter()
 router.register(
     r"programs", LearningPathAsProgramViewSet, basename="learning-path-as-program"
 )
 
-urlpatterns = router.urls
+urlpatterns = router.urls + [
+    path(
+        "learning-paths/<int:learning_path_id>/progress/",
```
Comment (Agrendalath) [inline learning_paths/api/v1/urls.py:18]: Shouldn't we use UUIDs here to hide the number of existing learning paths?

File: learning_paths/api/v1/utils.py [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+        data = response.json()
+    except HTTPError as err:
+        raise APIException(f"Error fetching completion for course {course_id}: {err}")
+
+    if data and data.get("results"):
+        return data["results"][0]["completion"]["percent"]
+    return 0.0
+
+
+def get_aggregate_progress(request, learning_path):
```
Comment (Agrendalath) [inline learning_paths/api/v1/utils.py:30]: It would make more sense to pass the username here instead of the whole request.

File: learning_paths/api/v1/utils.py [hunk 6]
```diff
@@ -0,0 +1,51 @@
...
+    """
+    Calculate the aggregate progress for all courses in the learning path.
+    """
+    steps = LearningPathStep.objects.filter(learning_path=learning_path)
+    total_courses = steps.count()
```
Comment (Agrendalath) [inline learning_paths/api/v1/utils.py:35]: Tiny nit: using `.count()` before iterating over the queryset adds an extra DB query. To avoid this, you can either:
1. Use `len()` here. This will fetch the queryset, so that it will be ready for the iteration.
2. Use either `len()` or `.count()` after creating the queryset iterator (inside or after the `for step in steps` loop) because it would use the cached result.

**Review by Agrendalath — CHANGES_REQUESTED** [review]
@pkulkark, the API works mostly correctly. Please see my comments and add unit tests for this code to verify the edge cases (we don't need e2e tests for the completion aggregator yet). Also, could you please take a look at the failing CI? I don't think we will deploy this in Palm, so we can target Redwood and use the Python 3.11 - Django 4.2 combination in our tests.

File: learning_paths/api/v1/utils.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+    lms_base_url = settings.LMS_ROOT_URL
+    username = request.user.username
+    completion_url = f"{lms_base_url}/completion-aggregator/v1/course/{course_id}/?username={username}"
+
+    try:
+        response = client.get(completion_url)
+        response.raise_for_status()
+        data = response.json()
+    except HTTPError as err:
+        raise APIException(f"Error fetching completion for course {course_id}: {err}")
```
[the remark this answers is not in the document]
Comment (pkulkark) [reply learning_paths/api/v1/utils.py]: Added.

File: learning_paths/api/v1/utils.py [hunk 8]
```diff
@@ -0,0 +1,51 @@
+import requests
+from django.conf import settings
+from openedx.core.djangoapps.catalog.utils import get_catalog_api_client
+from rest_framework.exceptions import APIException
+from ...models import LearningPathStep
+from requests.exceptions import HTTPError
+
+
+def get_course_completion(request, course_key, client):
```
[the remark this answers is not in the document]
Comment (pkulkark) [reply learning_paths/api/v1/utils.py]: Added.

File: README.rst [hunk 9]
```diff
@@ -15,3 +15,45 @@ The code in this repository is licensed under the Not open source unless
...
+
+Installation and Configuration
+******************************
+1. **Clone the Repository**
+
+    Clone the repository containing the plugin to the `src` directory under your devstack root:
+
+    ```bash
+    git clone <repository_url> <devstack_root>/src/learning-paths-plugin
+    ```
```
[the remark this answers is not in the document]
Comment (pkulkark) [reply README.rst]: Ah my bad. Corrected it.

File: learning_paths/api/v1/views.py (hunk 3 above)
Comment (pkulkark) [reply learning_paths/api/v1/views.py]: Updated it to use APIView with the permission class.

File: learning_paths/api/v1/urls.py (hunk 4 above)
Comment (pkulkark) [reply learning_paths/api/v1/urls.py]: Updated it to use uuid instead.

File: learning_paths/api/v1/utils.py (hunk 5 above)
Comment (pkulkark) [reply learning_paths/api/v1/utils.py:45]: Done.

File: learning_paths/api/v1/utils.py (hunk 6 above)
Comment (pkulkark) [reply learning_paths/api/v1/utils.py]: Good point. Updated it.

File: requirements/base.in [hunk 10]
```diff
@@ -8,3 +8,4 @@ djangorestframework
 edx-django-utils
 edx-opaque-keys
 openedx-atlas
+openedx-completion-aggregator
```
Comment (Agrendalath) [inline requirements/base.in:11]: Nit: please add a comment explaining why we need this dependency.

File: learning_paths/api/v1/tests/test_views.py [hunk 11]
```diff
@@ -29,3 +37,34 @@ def test_list_learning_paths_as_programs(self):
...
+        self.user = UserFactory()
+        self.client.force_authenticate(user=self.user)
+        self.learning_path = LearnerPathwayFactory.create()
+
+    @patch("learning_paths.api.v1.views.get_aggregate_progress")
+    def test_learning_path_progress_success(self, mock_get_aggregate_progress):
+        """
+        Test retrieving progress for a learning path.
+        """
+        mock_get_aggregate_progress.return_value = 0.75
```
Comment (Agrendalath) [inline learning_paths/api/v1/tests/test_views.py:54]: TIny nit: `return_value=0.75` can be specified directly in the `patch` decorator.

File: README.rst [hunk 12]
```diff
@@ -15,3 +15,40 @@ The code in this repository is licensed under the Not open source unless
 otherwise noted.
 
 Please see `LICENSE.txt <LICENSE.txt>`_ for details.
+
+Installation and Configuration
```
Comment (Agrendalath) [inline README.rst:19]: I forgot about one thing - [this part](https://github.com/open-craft/openedx-completion-aggregator/?tab=readme-ov-file#synchronous-vs-asynchronous-calculations) is essential to read before enabling this in a production environment, as installing this package can lead to an outage. We should probably use [a caution highlight](https://github.com/orgs/community/discussions/16925) in this README to make this more visible.

**Review by Agrendalath — APPROVED** [review]
> [!WARNING]  
> @pkulkark, please bump the version and update the changelog before merging this.

:+1:

* [x] I tested this: checked the API endpoint
* [x] I read through the code
* [x] I checked for accessibility issues: n/a
* [x] Includes documentation

File: requirements/base.in (hunk 10 above)
Comment (pkulkark) [reply requirements/base.in]: Added.

File: README.rst (hunk 12 above)
Comment (pkulkark) [reply README.rst:19]: Good call. Added it.
